mountpoint-s3/src/prefetch/cached_feed.rs

```diff
@@ -0,0 +1,440 @@
+where
+    Client: ObjectClient + Send + Sync + 'static,
+    Cache: DataCache<CacheKey> + Send + Sync + 'static,
+    Runtime: Spawn,
+{
+    fn spawn_get_object_request(
+        &self,
+        bucket: &str,
+        key: &str,
+        if_match: ETag,
+        range: RequestRange,
+        _preferred_part_size: usize,
+    ) -> RequestTask<TaskError<Client>> {
+        let range = get_aligned_request_range(range, self.inner.cache.block_size());
+
+        let start = range.start();
+        let size = range.len();
+
+        let (part_queue, part_queue_producer) = unbounded_part_queue();
+        trace!(?range, "spawning request");
+
+        let request_task = {
+            let inner = self.inner.clone();
+            let bucket = bucket.to_owned();
+            let key = key.to_owned();
+            let etag = if_match.clone();
+            let span = debug_span!("prefetch", ?range);
+
+            async move {
+                inner
+                    .get_from_cache(&bucket, &key, etag, range, part_queue_producer)
+                    .await;
+            }
+            .instrument(span)
+        };
+
+        let task_handle = self.runtime.spawn_with_handle(request_task).unwrap();
+
+        RequestTask::from_handle(task_handle, size, start, part_queue)
+    }
+}
+
+fn get_aligned_request_range(range: RequestRange, block_size: u64) -> RequestRange {
+    let object_size = range.object_size();
+    let offset = range.start();
+    let preferred_size = range.len();
+
+    // If the request size is bigger than a block size we will try to align it to block boundaries.
+    let offset_in_part = offset % block_size;
+    let size = if offset_in_part != 0 {
+        // if the offset is not at the start of the part we will drain all the bytes from that part first
+        let remaining_in_part = block_size - offset_in_part;
+        preferred_size.min(remaining_in_part as usize)
+    } else {
+        // if the request size is smaller than the block size, just return the block size
+        if preferred_size < block_size as usize {
+            block_size as usize
+        } else {
+            // if it exceeds block boundaries, trim it to the block boundaries
+            let request_boundary = offset + preferred_size as u64;
+            let remainder = request_boundary % block_size;
+            if remainder != 0 {
+                preferred_size + (block_size - remainder) as usize
+            } else {
+                preferred_size
+            }
+        }
+    };
+    RequestRange::new(object_size, offset, size)
+}
+
+#[derive(Debug)]
+struct CachedPartFeedInner<Client, Cache> {
+    client: Arc<Client>,
+    cache: Arc<Cache>,
+}
+
+impl<Client, Cache> Clone for CachedPartFeedInner<Client, Cache> {
+    fn clone(&self) -> Self {
+        Self {
+            client: self.client.clone(),
+            cache: self.cache.clone(),
+        }
+    }
+}
```

Comment:
Why not using derive?

Reply:
Generic parameters. See https://stegosaurusdormant.com/understanding-derive-clone/